File: 0460-lfu-cache/0460-lfu-cache.java
class LFUCache {
    Map<Integer, int[]> cacheMap = new HashMap<>();
    Map<Integer, LinkedHashSet<Integer>> frequencyMap = new HashMap<>();
    int minf = 0;
    int capacity;
    public LFUCache(int capacity) {
        this.capacity = capacity;
    }
    
    public int get(int key) {
        if(!cacheMap.containsKey(key)){
            return -1;
        }
        int[] freqValue = cacheMap.get(key);
        int freq = freqValue[0];
        int value = freqValue[1];
        LinkedHashSet<Integer> freqSet = frequencyMap.get(freq);
        freqSet.remove(key);
        LinkedHashSet<Integer> newFreqSet = frequencyMap.getOrDefault(freq + 1, new LinkedHashSet<>());
        newFreqSet.add(key);
        if(freqSet.size() == 0 && minf == freq){
            minf++;
        }
        frequencyMap.put(freq, freqSet);
        frequencyMap.put(freq + 1, newFreqSet);

        freqValue[0] = freqValue[0] + 1;
        cacheMap.put(key, freqValue);
        return value;
    }
    
    public void put(int key, int value) {
        if(cacheMap.containsKey(key)){
            int[] freqValue = cacheMap.get(key);
            freqValue[1] = value;
            cacheMap.put(key, freqValue);
            get(key);
        } else {
            if(cacheMap.size() >= capacity){
                LinkedHashSet<Integer> lowFreq = frequencyMap.get(minf);
                int keyToRemove = lowFreq.iterator().next();
                lowFreq.remove(keyToRemove);
                frequencyMap.put(minf, lowFreq);
                cacheMap.remove(keyToRemove);
            }
            cacheMap.put(key, new int[]{1, value});
            LinkedHashSet<Integer> oneFreq = frequencyMap.getOrDefault(1, new LinkedHashSet<>());
            oneFreq.add(key);
            frequencyMap.put(1, oneFreq);
            minf = 1;
        }
    }
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 0460-lfu-cache/0460-lfu-cache.py
class LFUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.low = 1
        self.freqdict = collections.defaultdict(int)
        self.removedict = collections.defaultdict(list)
        self.val = collections.defaultdict(int)

    def get(self, key: int) -> int:
        if key in self.freqdict:
            freq = self.freqdict[key]
            self.freqdict[key] = freq + 1
            self.removedict[freq].remove(key)
            self.removedict[freq + 1].append(key)
            if freq == self.low and not self.removedict[freq] :
                self.low = freq + 1
            return self.val[key]
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if key in self.freqdict:
            freq = self.freqdict[key]
            self.freqdict[key] = freq + 1
            self.removedict[freq].remove(key)
            self.removedict[freq + 1].append(key)
            self.val[key] = value
            if freq == self.low and not self.removedict[freq] :
                self.low = freq + 1
            return
        elif len(self.freqdict) >= self.capacity:
            abortkey = self.removedict[self.low][0]
            self.removedict[self.low].remove(abortkey)
            self.freqdict.pop(abortkey)
            self.val.pop(abortkey)
        self.removedict[1].append(key)
        self.freqdict[key] = 1
        self.low = 1
        self.val[key] = value
        


# Your LFUCache object will be instantiated and called as such:
# obj = LFUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)